### ISSUE TITLE
Extremely long compilation time for Circe module

### ISSUE THREAD

**Issue by chuwy** [hseeberger/akka-http-json#47]
Does any one else experiences an issue with extremely long compilation time?
I created small project with akka-http-json (particularly Circe) to show up a problem: https://github.com/chuwy/compile-problem/blob/master/src/main/scala/me.chuwy/example/Routes.scala#L10

If I remove `CirceSupport` from code, it compiles in a few seconds, but with `import CirceSupport._` or `extends CirceSupport` it takes ~100s on this simple code chunk and >5min on more complicated. And all this even without actual decoding/encoding.

**Comment by hseeberger** [comment]
I noticed long compilation, too. Probably fue to format derivation. Maybe @travisbrown has some insights.

**Comment by chuwy** [comment]
Yes, I can confirm it happens only for circe. I tried json4s and there's no this problem. I also tried it with circe 0.3.0-SNAPSHOT and problem seems still there.

**Comment by travisbrown** [comment]
The Shapeless-powered fully-automatic generic derivation in circe can be very slow in some situations (see the last two items in [this list](https://github.com/travisbrown/circe#warnings-and-known-issues) for some discussion).

There's work happening in Shapeless 2.3.0 that aims to improve this, and [some changes](https://github.com/travisbrown/circe/pull/164) I'm working on independently in circe also seem to make the situation incrementally better.

For now unfortunately the best solution is to fall back to semi-automatically derived instances, which means two lines of boilerplate for every type you need to serialized / deserialize:

```scala
import io.circe.{ Decoder, Encoder }, io.circe.generic.semiauto.deriveFor

implicit val myTypeEncoder: Encoder[MyType] = deriveFor[MyType].encoder
implicit val myTypeDecoder: Decoder[MyType] = deriveFor[MyType].decoder
```

Note that the runtime performance of derived instances (using both `auto` and `semiauto`) should be practically indistinguishable from manually-defined instances (there are benchmarks in circe to track this).

**Comment by chuwy** [comment]
Thank you Travis.
@hseeberger I suppose the only thing we can do is to place warning about it somewhere?

**Comment by nettok** [comment]
When I upgraded to circe 0.4.0-RC1 the compilation time (for a very simple thing) went down from 1 minute 25 seconds to just 3 seconds. :+1:

**Comment by hseeberger** [comment]
Nice. Hoping for a Circe 0.4 release anytime soon.

**Comment by hseeberger** [comment]
I think this can be closed, because compilation time with Circe 0.4 is much better than before.
